hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/http/SolrHttpClient.java

```diff
@@ -0,0 +1,197 @@
+import org.apache.http.client.protocol.HttpClientContext;
+import org.apache.http.config.Lookup;
+import org.apache.http.config.RegistryBuilder;
+import org.apache.http.conn.ssl.NoopHostnameVerifier;
+import org.apache.http.conn.ssl.TrustStrategy;
+import org.apache.http.impl.auth.SPNegoSchemeFactory;
+import org.apache.http.impl.client.BasicCredentialsProvider;
+import org.apache.http.impl.client.CloseableHttpClient;
+import org.apache.http.impl.client.HttpClients;
+import org.apache.http.ssl.SSLContextBuilder;
+import org.apache.http.util.EntityUtils;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import javax.inject.Singleton;
+import java.io.IOException;
+import java.security.KeyManagementException;
+import java.security.KeyStoreException;
+import java.security.NoSuchAlgorithmException;
+import java.security.Principal;
+import java.security.cert.CertificateException;
+import java.security.cert.X509Certificate;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.List;
+
+/**
+ * This class is used to make http
+ * GET and POST requests to Apache Solr.
+ */
+@Singleton
+public class SolrHttpClient {
+  private static final Logger LOG =
+      LoggerFactory.getLogger(SolrHttpClient.class);
+
+  private static CloseableHttpClient httpClient;
+  private static HttpClientContext context;
+
+  public static CloseableHttpClient getCloseableHttpClient()
+      throws KeyManagementException, NoSuchAlgorithmException,
+      KeyStoreException {
+    if (null != httpClient) {
+      return httpClient;
+    }
+    try {
+      System.setProperty("sun.security.krb5.debug", "true");
+      System.setProperty("javax.security.auth.useSubjectCredsOnly", "false");
+      Lookup<AuthSchemeProvider> authSchemeRegistry =
+          RegistryBuilder.<AuthSchemeProvider>create()
+              .register(AuthSchemes.SPNEGO, new SPNegoSchemeFactory(true))
+              .build();
+      httpClient = HttpClients.custom()
+          .setSSLHostnameVerifier(
+              NoopHostnameVerifier.INSTANCE)
+          .setSSLContext(new SSLContextBuilder().loadTrustMaterial(
+              null, new TrustStrategy() {
+                public boolean isTrusted(X509Certificate[] arg0, String arg1)
+                    throws CertificateException {
+                  return true;
+                }
+              }).build())
+          .setDefaultAuthSchemeRegistry(authSchemeRegistry).build();
+    } catch (KeyManagementException kme) {
+      LOG.error("KeyManagementException in creating http client instance", kme);
+      throw kme;
+    } catch (NoSuchAlgorithmException nsae) {
+      LOG.error("NoSuchAlgorithmException in creating http client instance",
+          nsae);
+      throw nsae;
+    } catch (KeyStoreException kse) {
+      LOG.error("KeyStoreException in creating http client instance", kse);
+      throw kse;
+    }
+    return httpClient;
+  }
+
+  private static HttpClientContext getHttpClientContext() {
```

Comment:
We should make this a synchronized function.  There is race condition for context check.

Reply:
The global httpClient need be closed when this SolrHttpClient is going to close.